cli/pcluster/utils.py

```diff
@@ -311,16 +274,130 @@ def get_instance_vcpus(instance_type):
+        )
+
+
+def _get_instance_families_from_types(instance_types):
+    """Return a list of instance families represented by the given list of instance types."""
+    families = set()
+    for instance_type in instance_types:
+        match = re.search(r"^([a-z0-9\-]+).", instance_type)
+        if match:
+            families.add(match.group(1))
+        else:
+            LOGGER.debug("Unable to parse instance family for instance type {0}".format(instance_type))
+    return list(families)
+
+
+def _instance_types_and_families_are_supported(candidate_types_and_families, known_types_and_families):
+    """Return a boolean describing whether the given instance types and families are known to be supported."""
+    unknowns = [candidate for candidate in candidate_types_and_families if candidate not in known_types_and_families]
+    if unknowns:
+        LOGGER.debug("Found the following unknown instance types/families: {0}".format(" ".join(unknowns)))
+    return not unknowns
+
+
+def get_supported_batch_instance_types():
     """
-    Get supported instance types.
+    Get the instance types supported by Batch in the desired region.
 
-    :return: the list of supported instance types
+    This is done by calling Batch's CreateComputeEnvironment with a bad
+    instance type and parsing the error message.
     """
+    supported_instance_types = get_supported_instance_types()
+    supported_instance_families = _get_instance_families_from_types(supported_instance_types)
+    supported_instance_types_and_families = supported_instance_types + supported_instance_families
     try:
-        instances = _get_json_from_s3(get_region(), "instances/instances.json")
-        return instances.keys()
-    except (ValueError, ClientError):
-        error("Unable to retrieve the list of supported instance types.")
+        emsg = _get_supported_instance_types_create_compute_environment_error_message()
+        parsed_instance_types_and_families = _parse_supported_instance_types_and_families_from_cce_emsg(emsg)
+        if _instance_types_and_families_are_supported(
+            parsed_instance_types_and_families, supported_instance_types_and_families
+        ):
+            supported_batch_types = parsed_instance_types_and_families
+        else:
+            supported_batch_types = supported_instance_types_and_families
+    except Exception as exception:
+        # When the instance types supported by Batch can't be parsed from an error message,
+        # log the reason for the failure and return instead a list of all instance types
+        # supported in the region.
+        LOGGER.debug(
+            "Failed to parse supported Batch instance types from a CreateComputeEnvironment "
+            "error message: {0}".format(exception)
+        )
+        supported_batch_types = supported_instance_types_and_families
+    return supported_batch_types
 
 
 def get_supported_compute_instance_types(scheduler):
```

Comment:
It seems to me we have now a big number of functions related to the instance_type retrieval, what about moving all of them into a `instance_type_utils.py` or something like this, to start cleaning up the mix we have in the `utils.py` module.

Reply:
i did consider this. the reason i didn't is because in my opinion it's only worth making a change like that if we make a utils directory containing various utility submodules (e.g. `utils.instance_types`).  that's a much bigger change than i'd like to make as part of this patch. i don't think it's best to only do it for utilities related to instance types.